### PR TITLE
Task/fix sorting of aggregate report table columns

### PR DESCRIPTION
**Pull request by esotrope**

### CODE REVIEW

File: webapp/src/main/webapp/src/app/aggregate-report/query-forms/claim-report-form.component.html [hunk 1]
```diff
@@ -574,7 +574,6 @@ <h3>{{'aggregate-report-form.section.preview-heading' | translate}}</h3>
           [options]="previewTable.options"
           [subjectDefinition]="subjectDefinition"
           [reportType]="previewTable.reportType"
-          [assessmentType]="previewTable.assessmentDefinition.typeCode"
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/aggregate-report/query-forms/claim-report-form.component.html]: turns out subjectDefinition has this already so this is not needed

File: webapp/src/main/webapp/src/app/aggregate-report/results/aggregate-report-table.component.ts [hunk 2]
```diff
@@ -648,13 +663,15 @@ export class AggregateReportTableComponent implements OnInit {
...
     const {
-      _identityColumnComparators: comparators,
       columns,
       identityColumns,
+      _identityColumnComparators,
       _orderingByColumnField,
       rows
     } = this;
 
+    const comparators = _identityColumnComparators.slice();
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/aggregate-report/results/aggregate-report-table.component.ts]: the sorting fix was to not modify the original comparators

File: webapp/src/main/webapp/src/app/aggregate-report/results/aggregate-report-table.component.ts [hunk 3]
```diff
@@ -754,25 +771,14 @@ export class AggregateReportTableComponent implements OnInit {
...
-
-    const { columns, identityColumns } = this;
-
-    // Assumes the ordered columns always start from the first column and extend to some terminal column
-    const comparator = ordering(ranking(identityColumns)).on((column: Column) => column.id).compare;
-    const orderedColumns: Column[] = columns.slice(0, identityColumns.length).sort(comparator);
-
-    columns.splice(0, identityColumns.length, ...orderedColumns);
-    this.sortRows(this.rows);
-  }
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/aggregate-report/results/aggregate-report-table.component.ts]: refactored to be more functional

File: webapp/src/main/webapp/src/app/aggregate-report/results/aggregate-report.component.html [hunk 4]
```diff
@@ -196,9 +196,8 @@ <h3 class="h3 label-group green pull-left">
         <div class="mt-sm">
           <aggregate-report-table #reportTable
                                   [options]="view.options"
-                                  [subjectDefinition]="getSubjectDefinitionForView(view)"
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/aggregate-report/results/aggregate-report.component.html]: this should improve performance slightly which is nice

File: webapp/src/main/webapp/src/app/aggregate-report/results/aggregate-report.component.ts [hunk 5]
```diff
@@ -329,18 +322,32 @@ export class AggregateReportComponent implements OnInit, OnDestroy {
...
+      effectiveReportType: reportType,
+      _subjectDefinitions,
+      assessmentDefinition,
+      columnOrderableItemProvider,
+      longitudinalDisplayType,
+      itemMapper,
+      reportViews,
+      chartMapper,
+      _viewComparator
+    } = this;
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/aggregate-report/results/aggregate-report.component.ts]: brought local dependencies to the top to make the method more sliceable later

File: webapp/src/main/webapp/src/app/aggregate-report/results/aggregate-target-overview.component.ts [hunk 6]
```diff
@@ -9,9 +8,6 @@ import { AggregateReportItem } from "./aggregate-report-item";
 export class AggregateTargetOverviewComponent {
 
   @Input()
-  set table(value: AggregateReportTable) {
-    this.overallRow = value.rows.find(row => row.subgroup.dimensionGroups[0].type === "Overall");
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/aggregate-report/results/aggregate-target-overview.component.ts]: demanded a lot more data than it used

File: webapp/src/main/webapp/src/app/aggregate-report/query-forms/claim-report-form.component.html (hunk 1 above)
Comment (malaffoon) [reply webapp/src/main/webapp/src/app/aggregate-report/query-forms/claim-report-form.component.html]: I'm glad that's so but `SubjectDefinition` here is a little confusing. Logically, a subject definition includes metadata about all the assessment types for a subject. Sounds like it is really SubjectAssessmentDefinition. But, your point is still valid.

Comment (esotrope) [reply webapp/src/main/webapp/src/app/aggregate-report/query-forms/claim-report-form.component.html]: good point - i'll add a todo to consider renaming it

File: webapp/src/main/webapp/src/app/aggregate-report/results/aggregate-report-table.component.ts (hunk 2 above)
Comment (malaffoon) [reply webapp/src/main/webapp/src/app/aggregate-report/results/aggregate-report-table.component.ts]: After null pointer problems, this has to be the most common coding issue.

File: webapp/src/main/webapp/src/app/aggregate-report/results/aggregate-report.component.ts (hunk 5 above)
Comment (malaffoon) [reply webapp/src/main/webapp/src/app/aggregate-report/results/aggregate-report.component.ts]: That's an interesting syntax, what's it doing exactly? The ` = this` part?

Comment (esotrope) [reply webapp/src/main/webapp/src/app/aggregate-report/results/aggregate-report.component.ts]: welcome to es6 - this is called a deconstructor.
below are functional equivalents:
```
const { a } = this;
const a = this.a;

const { a: b } = this;
const b = this.a;

const { a = {} } = this;
const a = this.a || {};
```
allows for more concise code - kind of like a `with` statement actually now that i think about it lol

**Pull request merged**